chalk-solve/src/clauses/syntactic_eq.rs

```diff
@@ -0,0 +1,165 @@
+use std::{iter, mem::replace};
+
+use chalk_engine::fallible::Fallible;
+use chalk_ir::{
+    cast::Cast,
+    fold::{shift::Shift, Fold, Folder, SuperFold},
+    interner::Interner,
+    AliasEq, AliasTy, Binders, BoundVar, DebruijnIndex, Goal, GoalData, Goals, ParameterKind,
+    ParameterKinds, ProgramClause, ProgramClauseData, ProgramClauseImplication, QuantifierKind, Ty,
+    TyData,
+};
+
+pub fn syn_eq_lower<I: Interner, T: Fold<I>>(interner: &I, clause: &T) -> <T as Fold<I>>::Result {
+    let mut folder = SynEqFolder {
+        interner,
+        new_params: vec![],
+        new_goals: vec![],
+        binders_len: 0,
+    };
+
+    clause
+        .fold_with(&mut folder, DebruijnIndex::INNERMOST)
+        .unwrap()
+}
+
+struct SynEqFolder<'i, I: Interner> {
+    interner: &'i I,
+    new_params: Vec<ParameterKind<()>>,
+    new_goals: Vec<Goal<I>>,
+    binders_len: usize,
+}
+
+impl<'i, I: Interner> Folder<'i, I> for SynEqFolder<'i, I> {
+    fn as_dyn(&mut self) -> &mut dyn Folder<'i, I> {
+        self
+    }
+
+    fn fold_ty(&mut self, ty: &Ty<I>, outer_binder: DebruijnIndex) -> Fallible<Ty<I>> {
+        let interner = self.interner;
+        let bound_var = BoundVar::new(DebruijnIndex::INNERMOST, self.binders_len);
+
+        let new_ty = TyData::BoundVar(bound_var).intern(interner);
+        match ty.data(interner) {
+            TyData::Alias(alias @ AliasTy::Projection(_)) => {
+                self.new_params.push(ParameterKind::Ty(()));
+                self.new_goals.push(
+                    AliasEq {
+                        alias: alias.clone(),
+                        ty: new_ty.clone(),
+                    }
+                    .cast(interner),
+                );
+                self.binders_len += 1;
+                ty.super_fold_with(self, outer_binder)?;
+                Ok(new_ty)
+            }
+            TyData::Function(_) => Ok(ty.clone()),
+            _ => Ok(ty.super_fold_with(self, outer_binder)?),
+        }
+    }
+
+    fn fold_program_clause(
+        &mut self,
+        clause: &ProgramClause<I>,
+        outer_binder: DebruijnIndex,
+    ) -> Fallible<ProgramClause<I>> {
+        let interner = self.interner;
```

Comment:
I think we should either 

* save the old values of `new_params`/`new_goals`, and restore before we return
* or assert that they are empty (if that is known to be true)

I don't see why they would be known to be empty, though, so I think we should save the values of the vector. The alternative would be to rewrite the code to not touch pre-existing things and just save the lengths.

Reply:
Probably we need to save `binders_len` too

Reply:
We know them to be empty because only domain goals and program clauses contain types (and EqGoal, but it is unique), and when folding domain goals we are guaranteed to not encounter another goal or clause.  If we restrict input types of `syn_lower` to only be `ProgramClause` and `Goal`, then the lowering should be correct.

Reply:
ok, then add assertions -- I guess it's true that we only have to worry about *domain goal*, and it cannot contain clauses. Right. I was thinking about how goals can contain clauses in general (and vice versa, of course).